src/aes.c

```diff
@@ -365,48 +365,129 @@ int aes_ecb_encrypt (unsigned char *out, const unsigned char *in, aes_context_t
                      const unsigned char *iv, aes_context_t *ctx) {
 
-  uint8_t tmp[AES_BLOCK_SIZE];
-  uint8_t old[AES_BLOCK_SIZE];
-  size_t i;
-  size_t n;
+  int n = in_len / 16; // number of blocks
+  in_len &= 15;        // remaining bytes
 
-  memcpy(tmp, iv, AES_BLOCK_SIZE);
+  __m128i ivec = _mm_loadu_si128((__m128i*)iv);
 
-  n = in_len / AES_BLOCK_SIZE;
-  for(i=0; i < n; i++) {
-    memcpy(old, &in[i * AES_BLOCK_SIZE], AES_BLOCK_SIZE);
-    aes_internal_decrypt(ctx, &in[i * AES_BLOCK_SIZE], &out[i * AES_BLOCK_SIZE]);
-    fix_xor(&out[i * AES_BLOCK_SIZE], tmp);
-    memcpy(tmp, old, AES_BLOCK_SIZE);
+  while(n >> 1) {      // equals: while(n > 1)
+    __m128i tmp1 = _mm_loadu_si128((__m128i*)in); in += 16;
+    __m128i tmp2 = _mm_loadu_si128((__m128i*)in); in += 16;
+
+
+    __m128i old_in1 = tmp1;
+    __m128i old_in2 = tmp2;
+
+    n--; n--;
+
+    tmp1 = _mm_xor_si128       (tmp1, ctx->rk_dec[ 0]); tmp2 = _mm_xor_si128       (tmp2, ctx->rk_dec[ 0]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 1]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 1]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 2]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 2]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 3]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 3]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 4]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 4]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 5]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 5]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 6]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 6]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 7]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 7]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 8]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 8]);
+    tmp1 = _mm_aesdec_si128    (tmp1, ctx->rk_dec[ 9]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[ 9]);
+
+    if(ctx->Nr > 10) {
+      tmp1 = _mm_aesdec_si128  (tmp1, ctx->rk_dec[10]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[10]);
+      tmp1 = _mm_aesdec_si128  (tmp1, ctx->rk_dec[11]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[11]);
+      if(ctx->Nr > 12) {
+        tmp1 = _mm_aesdec_si128(tmp1, ctx->rk_dec[12]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[12]);
+        tmp1 = _mm_aesdec_si128(tmp1, ctx->rk_dec[13]); tmp2 = _mm_aesdec_si128    (tmp2, ctx->rk_dec[13]);
+      }
+    }
+    tmp1 = _mm_aesdeclast_si128(tmp1, ctx->rk_enc[ 0]); tmp2 = _mm_aesdeclast_si128(tmp2, ctx->rk_enc[ 0]);
+
+    tmp1 = _mm_xor_si128 (tmp1, ivec); tmp2 = _mm_xor_si128 (tmp2, old_in1);
+
+    ivec = old_in2;
+
+    _mm_storeu_si128((__m128i*) out, tmp1); out += 16;
+    _mm_storeu_si128((__m128i*) out, tmp2); out += 16;
   }
 
-  return n * AES_BLOCK_SIZE;
+  if(n) {
+    __m128i tmp = _mm_loadu_si128((__m128i*)in);
+    __m128i old_in = tmp;
+    in += 16; n--;
+
+    tmp = _mm_xor_si128       (tmp, ctx->rk_dec[ 0]);
```

Comment:
and here

Reply:
Not here, this handles the last block which in turn renders the increments and decrements even completely superfluous. – Will remove.

Reply:
With the new commit (4 rails) the `if` still fits in better here as the following code is passed only once (2 or 3 blocks left) or never (1 or 0 blocks left).